collections/nemo_asr/nemo_asr/audio_preprocessing.py

```diff
@@ -0,0 +1,470 @@
+    """
+    @staticmethod
+    def create_ports():
+        input_ports = {
+            "input_signal": NeuralType({0: AxisType(BatchTag),
+                                        1: AxisType(TimeTag)}),
+
+            "length": NeuralType({0: AxisType(BatchTag)}),
+        }
+
+        output_ports = {
+            "processed_signal": NeuralType({0: AxisType(BatchTag),
+                                            1: AxisType(SpectrogramSignalTag),
+                                            2: AxisType(ProcessedTimeTag)}),
+
+            "processed_length": NeuralType({0: AxisType(BatchTag)})
+        }
+        return input_ports, output_ports
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        self.disable_casts = (self._opt_level == Optimization.mxprO1)
+
+        self.torch_windows = {
+            'hann': torch.hann_window,
+            'hamming': torch.hamming_window,
+            'blackman': torch.blackman_window,
+            'bartlett': torch.bartlett_window,
+            'none': None,
+        }
+
+    @torch.no_grad()
+    def forward(self, input_signal, length):
+        if self.disable_casts:
+            with amp.disable_casts():
+                processed_signal = self.get_features(
+                        input_signal.to(torch.float), length)
+        else:
+            processed_signal = self.get_features(input_signal, length)
+
+        processed_length = self.get_seq_len(length)
+        return processed_signal, processed_length
+
+    def get_features(self, input_signal, length):
+        # Called by forward(). Subclasses should implement this.
+        raise NotImplementedError
+
+    def get_seq_len(self, length):
+        # Called by forward(). Subclasses should implement this.
+        raise NotImplementedError
+
+
+class AudioToSpectrogramPreprocessor(AudioPreprocessor):
+    """Preprocessor that converts wavs to spectrograms.
+    Uses torchaudio's Spectrogram class as a featurizer.
+
+    Args:
+        sample_rate (int): Sample rate of the input audio data.
+            Defaults to 16000
+        window_size (float): Size of window for fft in seconds
+            Defaults to 0.02
+        window_stride (float): Stride of window for fft in seconds
+            Defaults to 0.01
+        n_fft (int): Length of FT window. If None, it uses the smallest power
+            of 2 that is larger than n_window_size.
+            Defaults to None
+        window (str): Windowing function for fft. can be one of ['hann',
+            'hamming', 'blackman', 'bartlett']
+            Defaults to "hamming"
+        normalized (bool): Whether to normalize by magnitude after stft
+        pad (int): Two sided padding
+    """
+    def __init__(
+            self, *,
+            sample_rate=16000,
+            window_size=0.02,
+            window_stride=0.01,
+            n_fft=None,
+            window="hamming",
+            normalized=True,
+            pad=8,
```

Comment:
Are you sure we want two-side padding here? This is audio padding, not spectrogram padding.
Also please be aware that torch.stft() has different defaults for padding than torchaudio.transforms.Spectrogram().
I personally would prefer the torch.stft() defaults over torchaudio.

Reply:
Removed pad option here according to discussion.
Turns out torchaudio calls torch.stft, and hard-codes the center arg to True anyway.